File: server/src/controllers/transactionAttachmentController.ts
import { Request, Response } from 'express';
import { query } from '../config/db.js';
import { v4 as uuidv4 } from 'uuid';
import { toMySQLDateTime } from '../utils/dataHelpers.js';
import type { AuthRequest } from '../middleware/auth.js';
import multer from 'multer';
import path from 'path';
import fs from 'fs';

// Helper function to get transaction attachment URL
function getTransactionAttachmentUrl(filename: string): string {
  const baseUrl = process.env.API_BASE_URL || 
                  process.env.SERVER_URL || 
                  (process.env.NODE_ENV === 'production' 
                    ? process.env.PRODUCTION_API_URL || 'https://your-api-domain.com'
                    : 'http://localhost:2222');
  
  return `${baseUrl}/uploads/transactions/${filename}`;
}

// Configure multer for transaction attachments
const transactionAttachmentsDir = path.join(process.cwd(), 'uploads', 'transactions');
if (!fs.existsSync(transactionAttachmentsDir)) {
  fs.mkdirSync(transactionAttachmentsDir, { recursive: true });
}

const transactionStorage = multer.diskStorage({
  destination: (req, file, cb) => {
    cb(null, transactionAttachmentsDir);
  },
  filename: (req, file, cb) => {
    const uniqueName = `${uuidv4()}${path.extname(file.originalname)}`;
    cb(null, uniqueName);
  },
});

export const uploadTransactionFiles = multer({
  storage: transactionStorage,
  limits: {
    fileSize: 5 * 1024 * 1024, // 5MB max per file (reduced from 10MB for better storage efficiency)
  },
});

// Get attachments for a transaction
export const getAttachments = async (req: Request, res: Response) => {
  try {
    const { transactionId } = req.query;

    if (!transactionId) {
      return res.status(400).json({ error: 'transactionId là bắt buộc' });
    }

    // Verify transaction exists
    const transaction = await query<any[]>(
      'SELECT id FROM material_transactions WHERE id = ?',
      [transactionId]
    );

    if (transaction.length === 0) {
      return res.status(404).json({ error: 'Không tìm thấy giao dịch' });
    }

    // Get attachments
    const attachments = await query<any[]>(
      'SELECT * FROM transaction_attachments WHERE transaction_id = ? ORDER BY created_at ASC',
      [transactionId]
    );

    // Convert attachments to full URLs
    const attachmentsWithUrls = attachments.map((att) => ({
      id: att.id,
      transactionId: att.transaction_id,
      filename: att.filename,
      originalFilename: att.original_filename,
      fileType: att.file_type,
      fileSize: att.file_size,
      fileUrl: getTransactionAttachmentUrl(att.filename),
      createdAt: att.created_at,
    }));

    res.json(attachmentsWithUrls);
  } catch (error: any) {
    console.error('Error fetching transaction attachments:', error);
    res.status(500).json({ error: 'Không thể lấy danh sách file đính kèm' });
  }
};

// Create attachments for a transaction
export const createAttachments = async (req: AuthRequest, res: Response) => {
  try {
    const userId = req.userId;
    if (!userId) {
      return res.status(401).json({ error: 'Không có quyền truy cập' });
    }

    const { transactionId } = req.body;

    if (!transactionId) {
      return res.status(400).json({ error: 'transactionId là bắt buộc' });
    }

    // Verify transaction exists
    const transaction = await query<any[]>(
      'SELECT id FROM material_transactions WHERE id = ?',
      [transactionId]
    );

    if (transaction.length === 0) {
      return res.status(404).json({ error: 'Không tìm thấy giao dịch' });
    }

    // Handle file attachments if any
    const attachments = [];
    if (req.files && Array.isArray(req.files) && req.files.length > 0) {
      const files = req.files as Express.Multer.File[];
      const createdAt = toMySQLDateTime();
      
      for (const file of files) {
        const attachmentId = uuidv4();
        const fileUrl = getTransactionAttachmentUrl(file.filename);

        await query(
          `INSERT INTO transaction_attachments (id, transaction_id, filename, original_filename, file_type, file_size, file_url, created_at)
           VALUES (?, ?, ?, ?, ?, ?, ?, ?)`,
          [
            attachmentId,
            transactionId,
            file.filename,
            file.originalname,
            file.mimetype,
            file.size,
            fileUrl,
            createdAt,
          ]
        );

        attachments.push({
          id: attachmentId,
          transactionId,
          filename: file.filename,
          originalFilename: file.originalname,
          fileType: file.mimetype,
          fileSize: file.size,
          fileUrl,
          createdAt,
        });
      }
    }

    res.status(201).json(attachments);
  } catch (error: any) {
    console.error('Error creating transaction attachments:', error);
    res.status(500).json({ error: 'Không thể tạo file đính kèm' });
  }
};

// Delete an attachment
export const deleteAttachment = async (req: AuthRequest, res: Response) => {
  try {
    const userId = req.userId;
    if (!userId) {
      return res.status(401).json({ error: 'Không có quyền truy cập' });
    }

    const { id } = req.params;

    // Get attachment to check if it exists
    const attachment = await query<any[]>(
      'SELECT * FROM transaction_attachments WHERE id = ?',
      [id]
    );

    if (attachment.length === 0) {
      return res.status(404).json({ error: 'Không tìm thấy file đính kèm' });
    }

    // Verify transaction exists (optional check)
    const transaction = await query<any[]>(
      'SELECT id FROM material_transactions WHERE id = ?',
      [attachment[0].transaction_id]
    );

    if (transaction.length === 0) {
      return res.status(404).json({ error: 'Không tìm thấy giao dịch' });
    }

    // Delete file from disk
    try {
      const filePath = path.join(transactionAttachmentsDir, attachment[0].filename);
      if (fs.existsSync(filePath)) {
        fs.unlinkSync(filePath);
      }
    } catch (fileError: any) {
      console.error(`Error deleting file ${attachment[0].filename}:`, fileError);
      // Continue with database deletion even if file deletion fails
    }

    // Delete attachment from database
    await query('DELETE FROM transaction_attachments WHERE id = ?', [id]);

    res.status(204).send();
  } catch (error: any) {
    console.error('Error deleting transaction attachment:', error);
    res.status(500).json({ error: 'Không thể xóa file đính kèm' });
  }
};

File: server.net/Controllers/TransactionAttachmentController.cs
using ConstructOs.Server.Config;
using ConstructOs.Server.Middleware;
using ConstructOs.Server.Utils;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace ConstructOs.Server.Controllers;

[ApiController]
[Route("api/transaction-attachments")]
public class TransactionAttachmentController : ControllerBase
{
    private readonly IDatabaseService _db;
    private readonly IConfiguration _configuration;
    private readonly IWebHostEnvironment _environment;

    public TransactionAttachmentController(IDatabaseService db, IConfiguration configuration, IWebHostEnvironment environment)
    {
        _db = db;
        _configuration = configuration;
        _environment = environment;
    }

    private string GetTransactionAttachmentUrl(string? filename)
    {
        if (string.IsNullOrEmpty(filename)) return string.Empty;
        if (filename.StartsWith("http://") || filename.StartsWith("https://")) return filename;
        
        var baseUrl = _configuration["API_BASE_URL"] ?? 
                     _configuration["SERVER_URL"] ?? 
                     (_configuration["ASPNETCORE_ENVIRONMENT"] == "Production"
                        ? _configuration["PRODUCTION_API_URL"] ?? "https://your-api-domain.com"
                        : "http://localhost:2222");
        
        return $"{baseUrl}/uploads/transactions/{filename}";
    }

    [HttpGet]
    public async Task<IActionResult> GetAttachments([FromQuery] string? transactionId)
    {
        try
        {
            if (string.IsNullOrEmpty(transactionId))
            {
                return BadRequest(new { error = "transactionId là bắt buộc" });
            }

            var transaction = await _db.QueryAsync<dynamic>(
                "SELECT id FROM material_transactions WHERE id = ?",
                new[] { transactionId }
            );

            if (!transaction.Any())
            {
                return NotFound(new { error = "Không tìm thấy giao dịch" });
            }

            var attachments = await _db.QueryAsync<dynamic>(
                "SELECT * FROM transaction_attachments WHERE transaction_id = ? ORDER BY created_at ASC",
                new[] { transactionId }
            );

            var attachmentsWithUrls = attachments.Select(att => new
            {
                id = att.id,
                transactionId = att.transaction_id,
                filename = att.filename,
                originalFilename = att.original_filename,
                fileType = att.file_type,
                fileSize = att.file_size,
                fileUrl = GetTransactionAttachmentUrl(att.filename?.ToString()),
                createdAt = att.created_at
            }).ToList();

            return Ok(attachmentsWithUrls);
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Error fetching transaction attachments: {ex}");
            return StatusCode(500, new { error = "Không thể lấy danh sách file đính kèm" });
        }
    }

    [HttpPost]
    [Authorize]
    public async Task<IActionResult> CreateAttachments([FromForm] string transactionId, [FromForm] IFormFileCollection? files)
    {
        try
        {
            var userId = HttpContext.GetUserId();
            if (string.IsNullOrEmpty(userId))
            {
                return Unauthorized(new { error = "Không có quyền truy cập" });
            }

            if (string.IsNullOrEmpty(transactionId))
            {
                return BadRequest(new { error = "transactionId là bắt buộc" });
            }

            var transaction = await _db.QueryAsync<dynamic>(
                "SELECT id FROM material_transactions WHERE id = ?",
                new[] { transactionId }
            );

            if (!transaction.Any())
            {
                return NotFound(new { error = "Không tìm thấy giao dịch" });
            }

            if (files == null || files.Count == 0)
            {
                return BadRequest(new { error = "Phải có ít nhất một file" });
            }

            var createdAt = DataHelpers.ToMySQLDateTime();
            var attachments = new List<dynamic>();

            foreach (var file in files)
            {
                var attachmentId = Guid.NewGuid().ToString();
                var (filename, url) = await FileUploadHelper.HandleFileUpload(file, "transactions", _configuration, _environment);

                await _db.ExecuteAsync(
                    @"INSERT INTO transaction_attachments (id, transaction_id, filename, original_filename, file_type, file_size, file_url, created_at)
                      VALUES (?, ?, ?, ?, ?, ?, ?, ?)",
                    new object[]
                    {
                        attachmentId, transactionId, filename, file.FileName, file.ContentType, file.Length,
                        url, createdAt
                    }
                );

                attachments.Add(new
                {
                    id = attachmentId,
                    transactionId,
                    filename,
                    originalFilename = file.FileName,
                    fileType = file.ContentType,
                    fileSize = file.Length,
                    fileUrl = url,
                    createdAt
                });
            }

            return StatusCode(201, attachments);
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Error creating transaction attachments: {ex}");
            return StatusCode(500, new { error = "Không thể tạo file đính kèm" });
        }
    }

    [HttpDelete("{id}")]
    [Authorize]
    public async Task<IActionResult> DeleteAttachment(string id)
    {
        try
        {
            var userId = HttpContext.GetUserId();
            if (string.IsNullOrEmpty(userId))
            {
                return Unauthorized(new { error = "Không có quyền truy cập" });
            }

            var attachment = await _db.QueryAsync<dynamic>(
                "SELECT * FROM transaction_attachments WHERE id = ?",
                new[] { id }
            );

            if (!attachment.Any())
            {
                return NotFound(new { error = "Không tìm thấy file đính kèm" });
            }

            var transaction = await _db.QueryAsync<dynamic>(
                "SELECT id FROM material_transactions WHERE id = ?",
                new[] { attachment.First().transaction_id }
            );

            if (!transaction.Any())
            {
                return NotFound(new { error = "Không tìm thấy giao dịch" });
            }

            try
            {
                var filePath = Path.Combine(_environment.ContentRootPath, "uploads", "transactions", attachment.First().filename?.ToString() ?? "");
                if (System.IO.File.Exists(filePath))
                {
                    System.IO.File.Delete(filePath);
                }
            }
            catch (Exception fileEx)
            {
                Console.Error.WriteLine($"Error deleting file {attachment.First().filename}: {fileEx}");
            }

            await _db.ExecuteAsync("DELETE FROM transaction_attachments WHERE id = ?", new[] { id });
            return NoContent();
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Error deleting transaction attachment: {ex}");
            return StatusCode(500, new { error = "Không thể xóa file đính kèm" });
        }
    }
}

